lms/djangoapps/program_enrollments/api/v1/views.py

```diff
@@ -268,27 +318,106 @@ def post(self, request, *args, **kwargs):
                         status.HTTP_422_UNPROCESSABLE_ENTITY
                     )
 
-        for enrollment_serializer in enrollments_to_create.values():
-            # create the model
+        # TODO: make this a bulk save - https://openedx.atlassian.net/browse/EDUCATOR-4305
+        for (student_key, _), enrollment_serializer in enrollments_to_create.items():
             enrollment_serializer.save()
-            # TODO: make this a bulk save
 
-        if not enrollments_to_create:
-            return Response(
-                status=status.HTTP_422_UNPROCESSABLE_ENTITY,
-                data=response_data,
-                content_type='application/json',
-            )
+        return self._get_created_or_updated_response(request, enrollments_to_create, response_data)
 
-        if len(request.data) != len(enrollments_to_create):
+    @verify_program_exists
+    def patch(self, request, **kwargs):
+        """
+        Modify the program enrollments for a list of learners
+        """
+        if len(request.data) > 25:
             return Response(
-                status=status.HTTP_207_MULTI_STATUS,
-                data=response_data,
+                status=status.HTTP_413_REQUEST_ENTITY_TOO_LARGE,
                 content_type='application/json',
             )
 
+        program_uuid = kwargs['program_uuid']
+        student_data = self._request_data_by_student_key(request, program_uuid)
+
+        response_data = {}
+        response_data.update(self._remove_duplicate_entries(request, student_data))
+
+        existing_enrollments = {
+            enrollment.external_user_key: enrollment
+            for enrollment in
+            ProgramEnrollment.bulk_read_by_student_key(program_uuid, student_data)
+        }
+
+        enrollments_to_create = {}
+
+        for external_user_key in student_data.keys():
+            if external_user_key not in existing_enrollments:
+                student_data.pop(external_user_key)
+                response_data[external_user_key] = CourseEnrollmentResponseStatuses.NOT_IN_PROGRAM
+
+        for external_user_key, enrollment in existing_enrollments.items():
+            data = {key: value for key, value in student_data[external_user_key].items() if key == 'status'}
```

Comment:
NIT: there's probably a better variable name than "data" (yeah, yeah, I know I wrote this part).

Reply:
and done